internal/server/http/dump.go

```diff
@@ -0,0 +1,285 @@
+	buf = append(buf, `" `...)
+	buf = append(buf, strconv.Itoa(status)...)
+	buf = append(buf, " "...)
+	buf = append(buf, strconv.Itoa(size)...)
+	return buf
+}
+
+// copied from gorilla/handlers and modified to add the body
+// writeLog writes a log entry for req to w in Apache Common Log Format.
+// ts is the timestamp with which the entry should be logged.
+// status and size are used to provide the response HTTP status and size.
+func writeLog(writer io.Writer, params handlers.LogFormatterParams, body string) {
+	buf := buildCommonLogLine(params.Request, params.URL, params.TimeStamp, params.StatusCode, params.Size)
+	// Append body if present
+	if len(body) > 0 {
+		buf = append(buf, " "...)
+		buf = append(buf, body...)
+	}
+	buf = append(buf, '\n')
+	writer.Write(buf)
+}
+
+// isBinaryContent checks to see if the body is a common binary content type
+func isBinaryContent(r *http.Request) bool {
+	contentType := r.Header.Get("Content-Type")
+	for _, binaryType := range binaryContentTypes {
+		if strings.HasPrefix(contentType, binaryType) {
+			return true
+		}
+	}
+	return false
+}
+
+func shouldRecordRequest(s *Server) bool {
+	return len(s.dumpRequestsPath) > 0 && s.dumpCh != nil
+}
+
+func getBody(r *http.Request, s *Server) string {
+	if s.logLevel == 0 && !shouldRecordRequest(s) {
+		return ""
+	}
+	bodyBytes, err := io.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("Error reading request body: %v\n", err)
+		return ""
+	}
+	r.Body = io.NopCloser(bytes.NewReader(bodyBytes)) // Reset the body
+	body := string(bodyBytes)
+	// if content is binary, encode it to base64
+	if isBinaryContent(r) {
+		body = base64.StdEncoding.EncodeToString(bodyBytes)
+	}
+	return body
+}
+
+// CustomLoggingHandler provides a way to supply a custom log formatter
+// while taking advantage of the mechanisms in this package
+func CustomLoggingHandler(out io.Writer, h http.Handler, s *Server) http.Handler {
+	return handlers.CustomLoggingHandler(out, h, func(writer io.Writer, params handlers.LogFormatterParams) {
+		body := getBody(params.Request, s)
+
+		var err error
+		if shouldRecordRequest(s) {
+			err = recordRequest(params.Request, s, body)
+		}
+
+		// log the request based on the log level
+		// if err is set, log the request, but only add the body if the log level is 2 or higher
+		if s.logLevel >= 2 {
+			writeLog(writer, params, body)
+		} else if err != nil || s.logLevel > 0 {
+			writeLog(writer, params, "")
+		}
+	})
+}
+
+func recordRequest(r *http.Request, s *Server, body string) error {
+	rd := getRequestData(r, body)
+	select {
+	case s.dumpCh <- rd:
+		// Successfully sent the request data to the channel
+	default:
+		// Handle the case where the channel is full
+		log.Println("Channel is full, dropping request and logging it instead:")
+		return fmt.Errorf("request dump channel is full")
+	}
+	return nil
+}
+
+// Goroutine function to write requests to a JSON file
+func RequestWriter(filePath string, requestChan <-chan *RequestData) {
```

Comment:
Looks like this function lacks proper control of its lifecycle, graceful shutdown, etc. Same for the channel, etc.

Reply:
Reworked this to have a context and a wait group to handle graceful shutdown.